### PR TITLE
Opensearch support

### PR DESCRIPTION
**Pull request by conorbros**

### CODE REVIEW

File: shotover-proxy/tests/opensearch_int_tests/mod.rs
Comment (rukai) [inline shotover-proxy/tests/opensearch_int_tests/mod.rs:74]: If this is a test case we want to get working eventually, lets not leave the code commented out. (otherwise it'll bit rot)
And instead just remove the failing assert and include a `TODO: add assert that X once Y functionality is fixed/implemented`

If this test case was a misstep and its not possible to get working then just delete it entirely.

File: shotover/src/codec/opensearch.rs
Comment (rukai) [inline shotover/src/codec/opensearch.rs:279]: same here, either:
* get this working
* leave out a failing assert + add TODO
* delete entirely

File: shotover/src/frame/opensearch.rs
Comment (rukai) [inline shotover/src/frame/opensearch.rs:18]: why is this an Arc?

File: test-helpers/src/docker_compose.rs
Comment (rukai) [inline test-helpers/src/docker_compose.rs:59]: bad merge

File: shotover/src/transforms/opensearch/mod.rs
Comment (rukai) [inline shotover/src/transforms/opensearch/mod.rs:168]: why cant we just use `crate::transforms::util::cluster_connection_pool::spawn_read_write_tasks`?

File: shotover/src/frame/opensearch.rs
Comment (rukai) [inline shotover/src/frame/opensearch.rs:27]: lets implement and use from_bytes instead of directly constructing via `new`?

File: shotover/src/frame/opensearch.rs
Comment (rukai) [inline shotover/src/frame/opensearch.rs:23]: this method is unused

File: shotover/src/codec/opensearch.rs
Comment (rukai) [inline shotover/src/codec/opensearch.rs:58]: considering this method returns a `Result`, all of these `TODO error handling` should just return an error instead of being left as a TODO.

File: shotover/src/sources/opensearch.rs
Comment (rukai) [inline shotover/src/sources/opensearch.rs:18]: either add an integration test for both opensearch Sink and Source TLS support or remove TLS support from this PR.

File: shotover/src/codec/opensearch.rs
Comment (rukai) [inline shotover/src/codec/opensearch.rs:174]: is this behavior you've encountered or something that you are guessing might happen?

File: shotover/src/codec/opensearch.rs
Comment (rukai) [inline shotover/src/codec/opensearch.rs:237]: what if we received a request that was a different version of HTTP?
Cant we use a crate for HTTP encoding?

File: shotover/src/codec/opensearch.rs
Comment (rukai) [inline shotover/src/codec/opensearch.rs:63]: We can merge these:
```rust
        let mut builder = Request::builder()
            .method(request.method.unwrap().to_owned().as_str())
            .uri(request.path.unwrap().to_owned().as_str());
```

File: shotover/src/codec/opensearch.rs
Comment (rukai) [inline shotover/src/codec/opensearch.rs:52]: this is equivalent to just  `let amt = match request.parse(src).unwrap() {`

Also, what is `amt`? Can you abbreviate it?

File: shotover/src/codec/opensearch.rs
Comment (rukai) [inline shotover/src/codec/opensearch.rs:64]: These can be just:
```rust
        let mut builder = Request::builder()
            .method(request.method.unwrap())
            .uri(request.path.unwrap());
```

File: shotover/src/codec/opensearch.rs
Comment (rukai) [inline shotover/src/codec/opensearch.rs:76]: These can be combined to avoid reallocating all of those strings.
```rust
        for header in request.headers {
            if header.name.is_empty() && header.value.is_empty() {
                break;
            }
            builder_headers.insert(
                HeaderName::from_bytes(header.name.as_bytes()).unwrap(),
                HeaderValue::from_bytes(header.value).unwrap(),
            );
        }
```

File: shotover-proxy/tests/opensearch_int_tests/mod.rs
Comment (rukai) [inline shotover-proxy/tests/opensearch_int_tests/mod.rs:199]: can we combine with test_create_index into a test_create_and_delete_index?
That way we are not at risk of breaking things by reordering or introducing new test cases.
e.g.
```rust
async fn test_create_and_delete_index(client: &OpenSearch) {
    client
        .indices()
        .create(IndicesCreateParts::Index("test-index"))
        .send()
        .await
        .unwrap();

    let exists_response = client
        .indices()
        .exists(IndicesExistsParts::Index(&["test-index"]))
        .send()
        .await
        .unwrap();

    assert_eq!(exists_response.status_code(), StatusCode::OK);
    
    client
        .indices()
        .delete(IndicesDeleteParts::Index(&["test-index"]))
        .send()
        .await
        .unwrap();

    let exists_response = client
        .indices()
        .exists(IndicesExistsParts::Index(&["test-index"]))
        .send()
        .await
        .unwrap();

    assert_eq!(exists_response.status_code(), StatusCode::NOT_FOUND);
}
```

File: shotover-proxy/tests/test-configs/opensearch-passthrough/docker-compose.yaml
Comment (rukai) [inline shotover-proxy/tests/test-configs/opensearch-passthrough/docker-compose.yaml:12]: our integration tests used to work without disabling authentication, what changed?

File: shotover/src/codec/opensearch.rs
Comment (conorbros) [reply shotover/src/codec/opensearch.rs:237]: We only accept HTTP/1 in the decode side of the codec. OpenSearch only started supporting HTTP/2 last year so I think it's safe to only support HTTP/1 for the moment.

File: shotover-proxy/tests/test-configs/opensearch-passthrough/docker-compose.yaml
Comment (conorbros) [reply shotover-proxy/tests/test-configs/opensearch-passthrough/docker-compose.yaml:12]: This disables SSL, which we don't have support for yet.

File: shotover/src/codec/opensearch.rs
Comment (conorbros) [reply shotover/src/codec/opensearch.rs:169]: Behaviour I encountered with OpenSearch

File: shotover/src/codec/opensearch.rs
Comment (rukai) [inline shotover/src/codec/opensearch.rs:112]: same feedback from request can be applied here

File: shotover-proxy/tests/test-configs/opensearch-passthrough/docker-compose.yaml
Comment (rukai) [reply shotover-proxy/tests/test-configs/opensearch-passthrough/docker-compose.yaml:12]: ahh, makes sense :+1:

File: test-helpers/src/docker_compose.rs
Comment (rukai) [inline test-helpers/src/docker_compose.rs:69]: and here

File: shotover/src/transforms/opensearch/mod.rs
Comment (conorbros) [reply shotover/src/transforms/opensearch/mod.rs:148]: Because HTTP is a synchronous protocol, we want to send a message, wait for a response and then send the next message.

Comment (rukai) [reply shotover/src/transforms/opensearch/mod.rs:148]: spawn_read_write_tasks should work for both synchronous and asynchronous protocols.
we use it for kafka which is synchronous.

File: shotover-proxy/tests/opensearch_int_tests/mod.rs
Comment (rukai) [inline shotover-proxy/tests/opensearch_int_tests/mod.rs:113]: These can all be: `assert!(results["took"].is_i64());`

File: shotover-proxy/tests/opensearch_int_tests/mod.rs
Comment (rukai) [inline shotover-proxy/tests/opensearch_int_tests/mod.rs:82]: If we assert that we get status ok, then the error reported to the developer is not very useful, all they know is that the status wasnt 200.
Instead, lets use a helper like this:
```rust
async fn assert_ok_and_get_json(response: Result<Response, Error>) -> Value {
    let response = response.unwrap();

    let status = response.status_code();
    let json = response.json::<Value>().await.unwrap();
    if status != StatusCode::OK {
        panic!("Opensearch query failed: {json:#?}");
    }
    json
}
```
Every query that we make to opensearch should use this helper, it will assert that the query succeeded, giving us the json if it did succeed.
If it fails then we get the error message from opensearch displayed in a panic.
e.g. I edited a query which was invalid, was I wrote this helper I got the following message which helped me understand what went wrong:

File: shotover-proxy/tests/opensearch_int_tests/mod.rs
Comment (rukai) [inline shotover-proxy/tests/opensearch_int_tests/mod.rs:17]: If we sort our search results then we can assert that we get the exact list of items that we expect:
```rust
pub async fn test_bulk(client: &OpenSearch) {
    assert_ok_and_get_json(
        client
            .indices()
            .create(IndicesCreateParts::Index("posts"))
            .send()
            .await,
    )
    .await;

    let mut body: Vec<BulkOperation<_>> = vec![];
    for i in 0..10 {
        let op = BulkOperation::index(json!({"title": "OpenSearch", "i": i}))
            .id(i.to_string())
            .into();
        body.push(op);
    }

    assert_ok_and_get_json(
        client
            .bulk(BulkParts::Index("posts"))
            .body(body)
            .refresh(Refresh::WaitFor)
            .send()
            .await,
    )
    .await;

    let results = assert_ok_and_get_json(
        client
            .search(SearchParts::None)
            .body(json!({
                "query": {
                    "match_all": {}
                },
                "sort": [
                    {
                        "i": {
                            "order": "asc"
                        }
                    }
                ]
            }))
            .allow_no_indices(true)
            .send()
            .await,
    )
    .await;

    assert!(results["took"].is_i64());
    let hits = results["hits"]["hits"].as_array().unwrap();
    assert_eq!(
        hits.iter().map(|x| &x["_source"]).collect::<Vec<_>>(),
        vec!(
            &json!({ "title": "OpenSearch", "i": 0 }),
            &json!({ "title": "OpenSearch", "i": 1 }),
            &json!({ "title": "OpenSearch", "i": 2 }),
            &json!({ "title": "OpenSearch", "i": 3 }),
            &json!({ "title": "OpenSearch", "i": 4 }),
            &json!({ "title": "OpenSearch", "i": 5 }),
            &json!({ "title": "OpenSearch", "i": 6 }),
            &json!({ "title": "OpenSearch", "i": 7 }),
            &json!({ "title": "OpenSearch", "i": 8 }),
            &json!({ "title": "OpenSearch", "i": 9 }),
        )
    );
}
```

File: shotover-proxy/tests/opensearch_int_tests/mod.rs
Comment (rukai) [inline shotover-proxy/tests/opensearch_int_tests/mod.rs:113]: a few improvements here:
```rust
    assert!(results["took"].is_i64());
    let hits = results["hits"]["hits"].as_array().unwrap();
    assert_eq!(
        hits.iter().map(|x| &x["_source"]).collect::<Vec<_>>(),
        vec!(&json!({
            "name": "John",
            "age": 30,
        }))
    );
    hits[0]["_id"].as_str().unwrap().to_owned()
```

* We can use `is_i64()` instead of `as_i64().is_some()`
* We can assert a list of expected `_sources` rather than just the length.
* We can use `["hits]` directly on a `Value` instead of calling `.as_object()`

File: shotover/src/codec/opensearch.rs
Comment (rukai) [inline shotover/src/codec/opensearch.rs:199]: this looks like a noop to me. 
If we reach this line then we are already in the ReadingBody state.

File: shotover/src/codec/opensearch.rs
Comment (rukai) [inline shotover/src/codec/opensearch.rs:50]: This should return a struct, its not clear what all the fields of this tuple mean.

File: shotover/src/codec/opensearch.rs
Comment (rukai) [reply shotover/src/codec/opensearch.rs:191]: This sounds like a bug to me, we should remove this hack and investigate whats going on.

File: shotover/src/codec/opensearch.rs
Comment (rukai) [inline shotover/src/codec/opensearch.rs:275]: this should go outside of the match so that it runs for both Encodable::Bytes and Encodable::Frame

File: shotover/src/codec/opensearch.rs
Comment (rukai) [reply shotover/src/codec/opensearch.rs:191]: I found that the issue occurs on:
```rust
        client
            .indices()
            .exists(IndicesExistsParts::Index(&["test-index"]))
            .send()
            .await,
```

And that if I do:
```rust
    assert_ok_and_get_json(
        client
            .indices()
            .exists(IndicesExistsParts::Index(&["test-index"]))
            .send()
            .await,
    )
    .await;
```

Then I get a panic in `assert_ok_and_get_json`:
```
thread 'opensearch_int_tests::passthrough_standard' panicked at 'called `Result::unwrap()` on an `Err` value: Error { kind: Http(reqwest::Error { kind: Decode, source: Error("EOF while parsing a value", line: 1, column: 0) }) }', shotover-proxy/tests/opensearch_int_tests/mod.rs:22:47
```

Both with and without shotover.
So seems to be a bug in either opensearch or the rust opensearch driver.

Comment (rukai) [reply shotover/src/codec/opensearch.rs:191]: Ah! figured it out.
The client sends an HTTP HEAD request
And as per https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/HEAD a head request, requests for just the headers without the body.
So we should change this to check for HEAD instead of `src.is_empty()`

File: shotover-proxy/tests/opensearch_int_tests/mod.rs
Comment (rukai) [reply shotover-proxy/tests/opensearch_int_tests/mod.rs:82]: Now that we know about `HEAD` requests, we should write this function like:
```rust
async fn assert_ok_and_get_json(response: Result<Response, Error>) -> Value {
    let response = response.unwrap();
    let status = response.status_code();

    if response.method() == Method::Head {
        if status != StatusCode::OK {
            panic!("Opensearch HEAD query returned status code {status}");
        }
        Value::Null
    } else {
        let json = response.json().await.unwrap();
        if status != StatusCode::OK {
            panic!("Opensearch query failed: {json:#?}");
        }
        json
    }
}
```

File: shotover/src/codec/opensearch.rs
Comment (rukai) [inline shotover/src/codec/opensearch.rs:116]: anyhow! already does interpolation, no need for the format macro.

**Comment by rukai** [comment]
Once we land this PR I think we should investigate using hyper instead of reimplementing so much http logic.
This example demonstrates a simple proxy with hyper https://github.com/hyperium/hyper/blob/master/examples/gateway.rs
We would be replacing our typical TcpCodecListener that we use in our Source implementations with `http1::Builder::new().serve_connection(io, service).await`.
And then our hyper service implementation would put the request in a shotover `Message` and send it down the transform chain.
I suspect that using hyper will gives us HTTP2 support for free.

I still think its a good idea to land this PR first though as it gives us a baseline to compare against.

File: shotover/src/codec/opensearch.rs
Comment (conorbros) [reply shotover/src/codec/opensearch.rs:191]: That would require some shared state between the encode and decode sides of the codecs, as the response to a HEAD request has nothing indicating it is that, it just responds the same as a GET request, but only the headers.

File: shotover-proxy/tests/opensearch_int_tests/mod.rs
Comment (rukai) [inline shotover-proxy/tests/opensearch_int_tests/mod.rs:162]: We should use `assert_ok_and_get_json` on all queries that we expect to succeed.

File: shotover/src/codec/opensearch.rs
Comment (rukai) [reply shotover/src/codec/opensearch.rs:202]: sounds like we'll need to add that shared state then.
checking `src.is_empty()` is no good as if we happen to perform the check just before receiving the bytes of the body we will ignore the body entirely.

Comment (conorbros) [reply shotover/src/codec/opensearch.rs]: Hmm, I had imagined that case would be very unlikely. Anyway the state wasn't difficult to implement just thought that it was better avoided if possible.

File: shotover/src/transforms/opensearch/mod.rs
Comment (conorbros) [reply shotover/src/transforms/opensearch/mod.rs:148]: Looking into this more, the current state of the sink is closer to `RedisSinkSingle` than the cluster implementation. I think keeping it simplest would be the best path forward for the time being.

Comment (rukai) [reply shotover/src/transforms/opensearch/mod.rs:148]: is this supposed to be a sink cluster or a single cluster?
We should name it either OpenSearchSinkSingle or OpenSearchSinkCluster to be consistent with our other sinks.

Comment (rukai) [reply shotover/src/transforms/opensearch/mod.rs:148]: I swapped this over to `spawn_read_write_tasks` and the tests still pass:
```diff
diff --git a/shotover/src/transforms/opensearch/mod.rs b/shotover/src/transforms/opensearch/mod.rs
index a91727a6..d7d8ca7d 100644
--- a/shotover/src/transforms/opensearch/mod.rs
+++ b/shotover/src/transforms/opensearch/mod.rs
@@ -1,13 +1,16 @@
-use crate::codec::{
-    opensearch::{OpenSearchCodecBuilder, OpenSearchDecoder, OpenSearchEncoder},
-    CodecBuilder, Direction,
-};
 use crate::message::Message;
 use crate::tcp;
 use crate::tls::AsyncStream;
 use crate::transforms::{
     Messages, Transform, TransformBuilder, TransformConfig, Transforms, Wrapper,
 };
+use crate::{
+    codec::{
+        opensearch::{OpenSearchCodecBuilder, OpenSearchDecoder, OpenSearchEncoder},
+        CodecBuilder, Direction,
+    },
+    transforms::util::cluster_connection_pool::spawn_read_write_tasks,
+};
 use anyhow::{anyhow, Result};
 use async_trait::async_trait;
 use futures::{FutureExt, SinkExt, StreamExt};
@@ -15,10 +18,13 @@ use serde::Deserialize;
 use std::pin::Pin;
 use std::time::Duration;
 use tokio::io::{ReadHalf, WriteHalf};
-use tokio::sync::mpsc;
+use tokio::sync::{mpsc, oneshot};
 use tokio_util::codec::{FramedRead, FramedWrite};
 use tracing::{trace, Instrument};
 
+use super::util::cluster_connection_pool::Connection;
+use super::util::Request;
+
 type PinStream = Pin<Box<dyn AsyncStream + Send + Sync>>;
 
 #[derive(Deserialize, Debug)]
@@ -96,22 +102,9 @@ impl Transform for OpenSearchSink {
             trace!("creating outbound connection {:?}", self.address);
 
             let tcp_stream = tcp::tcp_stream(self.connect_timeout, self.address.clone()).await?;
-            let stream = Box::pin(tcp_stream) as Pin<Box<dyn AsyncStream + Send + Sync>>;
-
-            let (decoder, encoder) = self.codec_builder.build();
-            let (stream_rx, stream_tx) = tokio::io::split(stream);
-            let outbound_tx = FramedWrite::new(stream_tx, encoder);
-            let outbound_rx = FramedRead::new(stream_rx, decoder);
-            let (response_messages_tx, response_messages_rx) = mpsc::unbounded_channel();
-
-            tokio::spawn(
-                server_response_processing_task(outbound_rx, response_messages_tx)
-                    .in_current_span(),
-            );
-            self.connection = Some(Connection {
-                response_messages_rx,
-                outbound_tx,
-            })
+            let (rx, tx) = tcp_stream.into_split();
+
+            self.connection = Some(spawn_read_write_tasks(&self.codec_builder, rx, tx));
         }
 
         let connection = self.connection.as_mut().unwrap();
@@ -120,61 +113,18 @@ impl Transform for OpenSearchSink {
 
         let mut result = Vec::with_capacity(messages_len);
         for message in requests_wrapper.requests {
+            let (tx, rx) = oneshot::channel();
             connection
-                .outbound_tx
-                .send(vec![message])
-                .await
-                .map_err(|err| {
-                    anyhow!("Failed to send messages to OpenSearch destination: {err:?}")
-                })?;
-
-            let message = connection
-                .response_messages_rx
-                .recv()
-                .await
-                .ok_or_else(|| anyhow!("Failed to receive message because OpenSearchSink response processing task is dead"))?;
+                .send(Request {
+                    message,
+                    return_chan: Some(tx),
+                })
+                .map_err(|_| anyhow!("Failed to send"))?;
+
+            let message = rx.await?.response?;
             result.push(message);
         }
 
         Ok(result)
     }
 }
-
-struct Connection {
-    outbound_tx: FramedWrite<WriteHalf<PinStream>, OpenSearchEncoder>,
-    response_messages_rx: mpsc::UnboundedReceiver<Message>,
-}
-
-async fn server_response_processing_task(
-    mut outbound_rx: FramedRead<ReadHalf<PinStream>, OpenSearchDecoder>,
-    response_messages_tx: mpsc::UnboundedSender<Message>,
-) {
-    loop {
-        tokio::select! {
-            responses = outbound_rx.next().fuse() => {
-                match responses {
-                    Some(Ok(messages)) => {
-                        for message in messages {
-                            if let Err(mpsc::error::SendError(_)) = response_messages_tx.send(message) {
-                                // RawSinkSingle dropped after a message was received from server,
-                                // request processor task shutting down
-                                return;
-                            }
-                        }
-                    }
-                    Some(Err(err)) => {
-                        tracing::error!("encountered error in OpenSearchSink stream: {err:?}");
-                        return;
-                    }
-                    None => {
-                        return;
-                    }
-                }
-            },
-            _ = response_messages_tx.closed() => {
-                // sink stream ended, shutdown the response processing task
-                return;
-            },
-        }
-    }
-}

```

Comment (conorbros) [reply shotover/src/transforms/opensearch/mod.rs:148]: I'm happy to switch it over to avoid adding more lines of code, I just found the initial one easier to reason about.

Comment (rukai) [reply shotover/src/transforms/opensearch/mod.rs:148]: yep I agree that spawn_read_write_tasks has a very complicated implementation. :/
In the future we should investigate if we can simplify it, I'm thinking we might be able to change it to use a single task without a performance hit.

In the meantime we are better off using a known good implementation rather than reimplementing spawn_read_write_tasks.
Either we do end up simplifying it and then all sinks benefit from the simplification or we dont because the complexity was actually important.

File: shotover/src/transforms/opensearch/mod.rs
Comment (rukai) [inline shotover/src/transforms/opensearch/mod.rs:26]: This is the name used by deserialization.
This needs to be OpenSearchSinkSingle.

File: shotover/src/transforms/opensearch/mod.rs
Comment (rukai) [inline shotover/src/transforms/opensearch/mod.rs:66]: We should also update this string as its user facing.

**Review by rukai — APPROVED** [review]
I think we are ready to land this.

File: shotover-proxy/tests/opensearch_int_tests/mod.rs
Comment (cjrolo) [inline shotover-proxy/tests/opensearch_int_tests/mod.rs:27]: We always get a valid json even if the status is not ok? Shouldn't we unwrap after testing the response status?

Comment (rukai) [reply shotover-proxy/tests/opensearch_int_tests/mod.rs:27]: Yep, opensearch returns valid json even when the http code is not successful.
We include the json in our panic message because it includes a detailed error message.

**Review by cjrolo — APPROVED** [review]
This was a long review, dived into some of the shotover internals, but for sure missed a lot of edges (and non-edges too!).  This should be more of a "non-binding approval" since I lack the knowledge about the internal system. BUT at the same time, it is a new part, so I get to know this one better.

**Review by cjrolo — APPROVED** [review]

**Comment by rukai** [comment]
Thanks!

**Pull request merged**
